video pipeline on rt10xx: Fix chicken-egg issue in init priority

ngphibang opened the pull request:
Similar to the mcxn947 (ov7670/smartDMA) issue, the CSI needs to be initialized BEFORE the camera sensor to provide clock to the camera sensor. However, the CSI node refers to the camera sensor node in the DT so it has to be initialized AFTER the sensor to be able to compile.

This is a **deadlock/chicken-egg** issue as Zephyr currently does not allow [circular dependency](https://github.com/zephyrproject-rtos/zephyr/issues/57708). Disable the init priority check as a work-around for this.

ngphibang committed:
drivers: video: mcux_mipi_csi2rx: Get data lanes number from devicetree

Get number of data lanes from device tree instead of hard-coding it.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
include: video: Add an utility function to get bytes per pixel

As getting bytes per pixel of a pixel format is a very common operation,
add an utility function for it instead of repeating the same codes in
different drivers.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

0xFarahFl committed:
drivers: video: mcux_mipi_csi2rx: Add set_ctrl callback

Add set_ctrl callback to propagate controls to the sensor.

Signed-off-by: Farah Fliss <farah.fliss@nxp.com>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

trunghieulenxp committed:
drivers: video: mcux_csi: Add support for changing frame rate

Add support for changing frame rate

Signed-off-by: Trung Hieu Le <trunghieu.le@nxp.com>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

trunghieulenxp committed:
drivers: video: mcux_mipi_csi2rx: Add support for changing frame rate

Add support for changing frame rate.

Signed-off-by: Trung Hieu Le <trunghieu.le@nxp.com>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: mcux_csi: Remove obsolete comment

Remove the obsolete comment about the init order of the CSI and the
camera sensors (e.g. mt9m114) which is not true anymore.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: mcux_csi: Add NXP copyright

The mcux csi is an NXP IP and the driver has been recently changed much
by NXP, so add NXP copyright to it.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

trunghieulenxp committed:
drivers: video: mcux_mipi_csi2rx: Set clocks according to pixel rate

Instead of fixing csi2rx clock frequencies, set them according to the
pixel rate got from the camera sensor.

Signed-off-by: Trung Hieu Le <trunghieu.le@nxp.com>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

trunghieulenxp committed:
drivers: video: mcux_csi: Fix start stop issue

When stopping, video_flush() moves all the remaining buffers
from the fifo_in to the fifo_out queue.

When restarting, we have two issues:
- There might be insufficient or no buffers in the fifo_in
queue to get the pipeline started.

- When dequeuing, users might immediatly get the incorrect
buffers in the fifo_out which do not come from interruption.

Fix it by moving all the remaining buffers from the fifo_out
to the fifo_in before starting the video capture.

Signed-off-by: Trung Hieu Le <trunghieu.le@nxp.com>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
dts-bindings: video: ov7725: Use video interfaces binding

Switch to use the new video interfaces binding

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
dts-bindings: video: ov5640: Use video interfaces binding

Switch to use the new video interfaces binding

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
dts-bindings: video: mipicsi2rx: Use video interfaces bindings

Switch to use the new video interfaces bindings

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
dts-bindings: video: csi: Use video interfaces bindings

Switch to use the new video interfaces bindings

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
boards: shields: nxp_btb44_ov5640: Add some endpoint properties

The ov5640 camera driver now supports both MIPI CSI2 (DPHY) and DVP
modes. It is in MIPI CSI2 mode in this overlay. Add bus-type property
for this. In this mode, data-lanes property is required as well.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
tests: drivers: build_all: video: Use the new endpoint binding

The csi, mipicsi2rx, ov5640 devices now use the new video interface
bindings. Change the test accordingly to take into account this.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
boards: shields: dvp_fpc24_mt9m114: Use the new endpoint bindings

As CSI now uses the new video-interfaces binding, this shield needs to
be changed too.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
We knew and did take the work-around internally but I think this should go into upstream so that people have an "out-of-the-box" working sample on i.MX RT10xx.

josuah reviewed (dismissed):
If I understand correctly, the init dependency is in the opposite direction than the `source = <&ov7670>;` suggests: phandles are considered as dependencies by the python scripts, whereas it is the opposite here.

If `source = <&ov7670>;` is replaced by `remote-endpoint-label = "ov7670_out";` (for instance), would that also be a workaround? In this case, there would not need to disable `CONFIG_CHECK_INIT_PRIORITIES` as there would be no more phandle at all?

Besides this, the workaround looks logical and avoids a compilation error, so seems necessary.

danieldegrasse commented:
> If I understand correctly, the init dependency is in the opposite direction than the `source = <&ov7670>;` suggests: phandles are considered as dependencies by the python scripts, whereas it is the opposite here.
> 
> If `source = <&ov7670>;` is replaced by `remote-endpoint-label = "ov7670_out";` (for instance), would that also be a workaround? In this case, there would not need to disable `CONFIG_CHECK_INIT_PRIORITIES` as there would be no more phandle at all?
> 
> Besides this, the workaround looks logical and avoids a compilation error, so seems necessary.

I think this would work, but I'm not sure Zephyr's DT tooling is capable of getting a device object via another method than a phandle. We could probably use something like `DT_NODELABEL(DT_STRING_TOKEN(remote_endpoint_label))` to get the node ID in this case? @decsny do you have any thoughts here?

josuah commented:
> We could probably use something like `DT_NODELABEL(DT_STRING_TOKEN(remote_endpoint_label))` to get the node ID in this case? @decsny do you have any thoughts here?

Absolutely! This is this PR and while I did not add ztests yet, this works as a workaround:
https://github.com/zephyrproject-rtos/zephyr/pull/73023/files#diff-d95b579605c12a2d00175b8b293dd3b30688ceed11edc95f473d4e3dc40b5babR3851

ngphibang commented:
To broaden the discussion scope, Linux has all the mechanism to parse and retrieve the peer (sub-)device for example with some `v4l2_async_notifier_parse_fwnode_endpoints_by_port()`.

And I think this circular dependency issue here is a sub-problem that we don't have "**async init**" in Zephyr where each device wait for each other to complete the initialization phase as each of them depends on others in different steps of initialization.

josuah commented:
So if I this gets merged first, it gives more time to design the DT endpoint API, and in the meantime the samples are working.

ngphibang committed:
drivers: video: mipi_csi2rx: Remove sensor device phandle

The peer remote device "sensor_dev" can be retrieved from the
remote-endpoint-label. Direct reference via phandle is not needed.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: csi: Remove source device phandle

The peer remote device "source_dev" can be retrieved from the
remote-endpoint-label. Direct reference via phandle is not needed.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: csi: Increase init priority

The CSI needs to be initialized BEFORE the camera sensor to provide
clock to the camera sensor. It is now possible to do so as direct
reference to the sensor via phandle is now not needed, so there
is no failed check on the init order anymore when compiling.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>
